Add scikit example notebook

schencej opened the pull request.

brianhhu commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "## Introduction\n",
+    "\n",
+    "The purpose of this example is to demonstrate the API's ability to work with a variety of machine learning libraries.\n",
+    "Here we use scikit learn's `LogisticRegression` class to implement a logistic regression model for image classification of the MNIST dataset.\n",
```
```suggestion
    "Here we use scikit-learn's `LogisticRegression` class to implement a logistic regression model for image classification of the MNIST dataset.\n",
```

brianhhu commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "from functools import reduce\n",
+    "from math import ceil\n",
+    "\n",
+    "def app(\n",
```
I think here it would be nice for the app to display the digit below each saliency map, similar to above for the scikit-learn based visualization.

brianhhu commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "    \n",
+    "    # Generate saliency maps\n",
+    "    sal_maps = []\n",
+    "    for i in range(num_imgs):    \n",
```
I think here, it would be nice to also visualize saliency maps for individual MNIST examples. We could make the story to be: 1) as a form of "local" explanation, show saliency maps for example digit exemplars, and 2) as a form of "global" explanation, show average saliency maps for each digit class.

Related to this, it is not surprising to me that the saliency maps agree with the coefficients learned by the logistic regression model since we are operating in the linear regime (i.e. each coefficient is a measure of influence on the model's final prediction, and since each coefficient maps to a certain pixel, this by definition maps directly to a pixel's influence on the final model prediction).

One thing that could be potentially interesting to try would be to borrow the MLP from the second scikit-learn [example](https://scikit-learn.org/stable/auto_examples/neural_networks/plot_mnist_filters.html#sphx-glr-auto-examples-neural-networks-plot-mnist-filters-py) (which is more than 1 layer) and apply our saliency maps there. The model is more complex and also achieves higher accuracy. I believe in that example they visualize the weights of the first layer, which are not great. This would be a slightly more involved example, but I still think instructive for the reader. @Purg Any thoughts on this?

Purg replied:
Adding a follow-on MLP take sounds like a good idea to me. In adding the MLP example this into *this* notebook, I don't think we need to reproduce the reference example, just refer to it (since their results are there). We can simply create the `MLPClassifier` instance, fit it, and apply the model as a blackbox to our "app" for saliency visualization.

Side note, even their example states that the MLP model will warn that it does not converge, so it doesn't surprise me that their visualization might not be that great (it's also just the first layer out of 50, so I'm not sure what they were expecting to get out of that visualization...). We can either duplicate their hyperparameters to provide a comparable result, or tweak the iteration/learning-rate to "better" the model for our visualization. I might suggest the former for comparability while noting that we understand that non-convergence is happening.

brianhhu commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+   "source": [
+    "## Blackbox Classifier\n",
+    "\n",
+    "Here we wrap our logistic regression in a `ClassifyImage` class to comply with the API's interface."
```
Might be useful to mention this as part of `smqtk-classifier`:

```suggestion
    "Here we wrap our logistic regression in smqtk-classifier's `ClassifyImage` class to comply with the API's interface."
```

brianhhu reviewed (commented):
In general, LGTM- made a few minor comments here and there. I also made some comments about going a bit deeper into the MNIST problem pending Paul's thoughts/feedback.

schencej commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from xaitk_saliency.impls.gen_image_classifier_blackbox_sal.rise import RISEStack\n",
```
Oops... I played around with the RISE algorithm for a while but the maps never looked very good. I'm not sure it's suitable for small images like what MNIST consists of. I forgot to remove the import here.

Purg commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "The purpose of this example is to demonstrate the API's ability to work with a variety of machine learning libraries.\n",
+    "Here we use scikit learn's `LogisticRegression` class to implement a logistic regression model for image classification of the MNIST dataset.\n",
+    "This is inspired and partially taken from thie example of scikit's website:<br>\n",
+    "&emsp;&emsp;https://scikit-learn.org/stable/auto_examples/linear_model/plot_sparse_logistic_regression_mnist.html\n"
```
I think it would be good to add more to this introduction, specifically about what we are adding to this example on top of the reference example from within scikit-learn.

For example, the example shows, effectively, an attention visualization, as we are duplicating here. The visualization of saliency heatmaps below seems very similar. I believe it is insinuated in the "Calling the Application" section, but we could be explicit here stating that the Scikit-Learn example visualization relies on *intrinsics* of the model in order to determine the "attention" visualization (whitebox), however our API's approach is classifier agnostic, "blackbox", and shows very comparable results without knowing anything about the specific classifier model being used, though likely some amount more processing intensive.

Purg commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "        for img in image_iter:\n",
+    "            images = np.vstack((images, img.reshape(1, -1)))\n",
+    "        \n",
+    "        return clf.decision_function(images)\n",
```
This technically does not fit the output data type of the interface API (returns iterator of dict instances). This seems to happen to work due to the classes being integers and the output elements index-able arrays. Maybe we don't need to change the code, but at least comment that the API is not being "met" due to special circumstances.

schencej replied:
I think I fixed this? No totally sure.

Purg replied:
Partially: it's now set to return an iterator, however the parent API states an iterator of dictionaries. Also, images are documented to be input as 2- or 3-dim. A more "correct" implementation might be:
```python
class MNISTClassifier (ClassifyImage):
    ...

    def classify_images(self, image_iter):
        # Yes, "images" in this example case are really 1-dim (28*28=784).
        # MLP input needs a (n_samples, n_features) matrix input.
        images = np.asarray(list(image_iter))  # may fail because input is not consistent in shape
        images = images.reshape(-1, 28*28)  # may fail because input was not the correct shape
        return (
            {idx: p_i for idx, p_i in zip(range(10), p)}
            for p in mlp.predict_proba(images)
        )
```

Purg commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+   "source": [
+    "## Heatmap Generation\n",
+    "\n",
+    "We create a `SlidingWindowStack` object to carry out our image perturbation and heatmap generation."
```
```suggestion
    "We create an instance of `SlidingWindowStack`, an implementation of the `GenerateImageClassifierBlackboxSaliency` interface, to carry out our image perturbation and heatmap generation."
```

Purg commented on examples/Scikit.ipynb:
```diff
@@ -0,0 +1,330 @@
+    "    for i in range(num_imgs):\n",
+    "        max_val = combined_maps[i].max()\n",
+    "        plt.subplot(2, ceil(num_imgs/2), i + 1)\n",
+    "        plt.imshow(combined_maps[i], cmap=plt.cm.RdBu, vmin=-max_val, vmax=max_val)\n",
```
I'm not sure the use of `vmin=-max_val, vmax=max_val` is appropriate here. The value is not symmetric in the heatmaps. The scikit-learn example uses this because it takes the global coef matrix absolute-value max.

We also know via API documentation that saliency heatmaps are constrained to return in the `[-1,1]` range. We may want to normalize the `combined_maps` by `len(sal_maps)` and then pass `-1, 1` to `vmin, vmax` respectively. @brianhhu please double check me on this.

This generates the following output:
![image](https://user-images.githubusercontent.com/735270/134057813-1c56833a-f899-4f75-98e9-d176d8c722f5.png)

Additionally may want to output a color bar to be more clear on color meanings, unless we are intending to assume the scikit-learn example output is well understood.

Purg reviewed (commented):
Similar to the change note comment, I would suggest being a little more descriptive in the name of the notebook file/title.

schencej commented:
I swapped out the logistic regression for the MLP and I think the narrative is much better. However, there is a lot more negative saliency in each map with this approach and the averaging makes a lot of the positive values hard to see when using the [-1,1] scale. Currently I only plot the positive saliency and I scale each individual map to span from [0, 1] in order to increase visibility. Let me know if you think this is appropriate.

Purg commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,392 @@
+   "source": [
+    "class MNISTClassifier (ClassifyImage):\n",
+    "    def get_labels(self):\n",
+    "        return [i for i in range(10)]\n",
```
```suggestion
    "        return list(range(10))\n",
```

Purg commented:
> ... However, there is a lot more negative saliency in each map with this approach and the averaging makes a lot of the positive values hard to see when using the [-1,1] scale. Currently I only plot the positive saliency and I scale each individual map to span from [0, 1] in order to increase visibility. Let me know if you think this is appropriate.

Saliency is not just about what "positively" contributes to a prediction: knowing about what a model down-weights can matter, too. I'll leave this more for @brianhhu to validate, but maybe in this case we are seeing that the trained MLP classifier here has learned less about what parts of the input "contributes" to a class prediction, but what *does not* contribute. E.g. it might be more predicting an input is not "zero" because there are active pixels in the center, not that there isn't input in a "zero" shape (if that makes sense written down...).

If @brianhhu also thinks this is a correct interpretation maybe we can use this as a teachable moment about the positive/negative saliency dichotomy and extend the narrative of this example.

schencej commented:
I added back in the regression content and reorganized a bit.

I also added a blurb in the "Calling the Application" section in the "Regression" part explaining the connection between the saliency map and the coefficients.

Please take a look.

brianhhu reviewed (commented):
LGTM, the narrative is much improved! I made a few suggested edits for clarification and readability.

brianhhu commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,544 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAj8AAAE6CAYAAAD5iTMZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAC0X0lEQVR4nOz9ebQkyZ7fBX5+Zu4eHsvd8uZaWXt1vVaXXi+8Vg9SN91qRjDDDKNGGsRBMEJo0IHhwLBIHIYRoIUBaY7gzKAZkEAggYTEJiTQiJGEAIlG6qfpbrVet/p191vqVb3asjIrM+/Nu8Tqi9n8YWYeHpH3Zt7c6i5hn3PuibgRHh7h5u5mX/ttJtZaIpFIJBKJRFYFddo/IBK ...
```
Can you check these output saliency maps? Somehow the saliency maps for all digits seemed to have collapsed into a single output, which is different from your previous commits. I couldn't find any obvious changes with the `app` function itself, so it may just need an output refresh.

schencej replied:
Not sure how I missed that... it's fixed now

schencej replied:
Not sure how I missed that... it's fixed now

brianhhu commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,544 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAj8AAAFCCAYAAAAe+Ly1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAut0lEQVR4nO3deZxkVX338e+vqrfp2fdhG4YgGNkx4oMgShSJD0YEJT6IAhMlxjzGKDE+xmgiWUxiElFfLtFAFJeAMRCJEYNGZUQRVwYUhbDOgAwMM8MMs/d0V5/nj3O67j23q3u6e6q6uu75vF+vevW599y1flXdvz7nnnvNOScAAIBUVNp9AAAAANOJ5AcAACSF5AcAACSF5AcAACS ...
```
Similarly, please check the output saliency maps here too. Thanks!

brianhhu commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
+# Example Data
+examples/scikit-learn-example/
```
Can you add this entry to the [.gitignore file within the examples directory](https://github.com/XAITK/xaitk-saliency/blob/master/examples/.gitignore) rather than the overall repo? i.e. add a new line like:

```
/scikit-learn-example/**
```

brianhhu reviewed (commented):
Please take a look at the minor review comments. It would also be nice to add Colab support by adding a simple header to the top of the notebook. Feel free to take a look at the other examples for how to do so. Thanks!

brianhhu commented on examples/.gitignore:
```diff
@@ -2,3 +2,4 @@
 /viame-example/**
 **/imagenet_classes.txt*
 /*.jpg
+/scikit-learn-exampel/**
```
`example` is mispelled, and can you move this up to where the other example directories are defined?

```suggestion
/scikit-learn-example/**
```

brianhhu reviewed (approved):
LGTM, I think technically only `xaitk-saliency` needs to be installed in the first cell (since it will pull in the other dependencies), but that's a minor issue.

Purg commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,572 @@
+    }
+   ],
+   "source": [
+    "!python -c \"import xaitk_saliency\" || pip install -q \"xaitk-saliency\""
```
This error maybe suppressed by adding a STDERR redirect in the lhs command:
```python
!python -c "import xaitk_saliency" 2>/dev/null || pip install -q "xaitk-saliency"
```

Of course, checking if it is installed is a little overkill as a simpler invocation of `!pip install -q xaitk-saliency` will be a no-op when the package is already installed.

brianhhu replied:
Do you know what is the cause of the output message? In my local poetry dev environment with jupyter notebook installed, running this cell is effectively a noop.

Purg replied:
The warning is the failure of the `python -c "import xaitk_saliency"` call, which is expected if the package is not installed. As mentioned above, we don't *need* this check since the pip install is idempotent.

Purg commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,572 @@
+    "# Split data into test and train sets\n",
+    "train_samples = 20000\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(\n",
+    "    X, y, train_size=train_samples, test_size=10000)\n",
```
I just realized that this could use a fixes random seed to result in consistent results from run to run
```suggestion
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y, train_size=train_samples, test_size=10000, random_state=0)\n",
```

Purg commented on examples/MNIST_scikit_saliency.ipynb:
```diff
@@ -0,0 +1,572 @@
+    "    X, y, train_size=train_samples, test_size=10000)\n",
+    "\n",
+    "# Define model, lower C value gives higher regulation\n",
+    "clf = LogisticRegression(\n",
+    "    C=50. / train_samples, penalty='l2', solver='saga', tol=0.1\n",
+    ")\n",
```
This could also use a random state seed
```suggestion
    "clf = LogisticRegression(\n",
    "    C=50. / train_samples, penalty='l2', solver='saga', tol=0.1, random_state=0\n",
    ")\n",
```

Purg reviewed (approved):
LGTM. The following comments are minor and should not block acceptance.

Purg commented:
Thanks for the updates! LGTM and merging
